File: app/src/main/java/com/coppermobile/mysampleusingmvvmlivedata/data/source/local/LocalDishDataSource.java
package com.coppermobile.mysampleusingmvvmlivedata.data.source.local;

import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;

import com.coppermobile.mysampleusingmvvmlivedata.data.Comments;
import com.coppermobile.mysampleusingmvvmlivedata.data.Dish;
import com.coppermobile.mysampleusingmvvmlivedata.data.source.DataSource;
import com.coppermobile.mysampleusingmvvmlivedata.utils.AppExecutors;

import java.util.List;


public class LocalDishDataSource implements DataSource<Dish> {

    private static volatile LocalDishDataSource INSTANCE;
    private DishDAO dishDAO;
    private AppExecutors appExecutors;

    private LocalDishDataSource(@NonNull AppExecutors appExecutors, @NonNull DishDAO dishDAO) {
        this.appExecutors = appExecutors;
        this.dishDAO = dishDAO;
    }

    public static LocalDishDataSource getInstance(@NonNull AppExecutors appExecutors, @NonNull DishDAO dishDAO) {
        if (INSTANCE == null) {
            synchronized (LocalDishDataSource.class) {
                if (INSTANCE == null) {
                    INSTANCE = new LocalDishDataSource(appExecutors, dishDAO);
                }
            }
        }
        return INSTANCE;
    }

    @Override
    public void insertData(Dish data) {

    }

    @Override
    public void insertData(List<Dish> data) {
        Runnable insertRunnable = () -> dishDAO.insertDishes(data);
        appExecutors.getDiskIO().execute(insertRunnable);
    }

    @Override
    public void deleteData(Dish data) {

    }

    @Override
    public LiveData<List<Comments>> getComments(int id) {
        return null;
    }

    @Override
    public LiveData<List<Dish>> getAllDishes() {
        return dishDAO.getAllDishes();
    }
}

File: app/src/main/java/com/coppermobile/kotlinsamplemvvmlivedataroom/data/source/local/LocalDishDataSource.kt
package com.coppermobile.kotlinsamplemvvmlivedataroom.data.source.local

import android.arch.lifecycle.LiveData
import com.coppermobile.kotlinsamplemvvmlivedataroom.data.Comments
import com.coppermobile.kotlinsamplemvvmlivedataroom.data.Dish
import com.coppermobile.kotlinsamplemvvmlivedataroom.data.source.DataSource
import com.coppermobile.kotlinsamplemvvmlivedataroom.utils.AppExecutors

class LocalDishDataSource private constructor(private val appExecutors: AppExecutors, private val dishDao: DishDao) : DataSource<Dish> {

    companion object {

        @Volatile
        private var INSTANCE: LocalDishDataSource? = null

        fun getInstance(appExecutors: AppExecutors, dishDao: DishDao): LocalDishDataSource {
            if (INSTANCE == null) {
                synchronized(this) {
                    INSTANCE = LocalDishDataSource(appExecutors, dishDao)
                }
            }
            return INSTANCE!!
        }
    }

    override fun insertData(data: Dish) {
    }

    override fun insertData(data: List<Dish>) {
        val insertRunnable = Runnable { dishDao.insertDishes(data) }
        appExecutors.diskIO.execute(insertRunnable)
    }

    override fun deleteData(data: Dish) {

    }

    override fun getComments(id: Int): LiveData<List<Comments>>? {
        return null
    }

    override fun getAllDishes(): LiveData<List<Dish>> {
        return dishDao.getAllDishes()
    }
}